lua/neogit/lib/git/repository.lua

```diff
@@ -45,20 +38,148 @@ M.create = function(_path)
     },
     rebase = {
       items = {},
-      head = "",
+      head = nil,
     },
     cherry_pick = {
       items = {},
-      head = "",
+      head = nil,
     },
     merge = {
       items = {},
       head = nil,
       msg = nil,
     },
   }
+end
+
+local meta = {
+  __index = function(self, method)
+    return self.state[method]
+  end,
+}
+
+local M = {}
+
+M.state = empty_state()
+M.lib = {}
+
+function M.reset(self)
+  self.state = empty_state()
+end
+
+function M.refresh(self, lib)
+  local refreshes = {}
+
+  if lib and type(lib) == "table" then
+    if lib.status then
+      self.lib.update_status(self.state)
+      a.util.scheduler()
+    end
+
+    if lib.branch_information then
+      table.insert(refreshes, function()
+        logger.debug("[REPO]: Refreshing branch information")
+        self.lib.update_branch_information(self.state)
+      end)
+    end
+
+    if lib.rebase then
+      table.insert(refreshes, function()
+        logger.debug("[REPO]: Refreshing rebase information")
+        self.lib.update_rebase_status(self.state)
+      end)
+    end
+
+    if lib.cherry_pick then
+      table.insert(refreshes, function()
+        logger.debug("[REPO]: Refreshing cherry-pick information")
+        self.lib.update_cherry_pick_status(self.state)
+      end)
+    end
+
+    if lib.merge then
+      table.insert(refreshes, function()
+        logger.debug("[REPO]: Refreshing merge information")
+        self.lib.update_merge_status(self.state)
+      end)
+    end
+
+    if lib.stashes then
+      table.insert(refreshes, function()
+        logger.debug("[REPO]: Refreshing stash")
+        self.lib.update_stashes(self.state)
+      end)
+    end
+
+    if lib.unpulled then
+      table.insert(refreshes, function()
+        logger.debug("[REPO]: Refreshing unpulled commits")
+        self.lib.update_unpulled(self.state)
+      end)
+    end
+
+    if lib.unmerged then
+      table.insert(refreshes, function()
+        logger.debug("[REPO]: Refreshing unpushed commits")
+        self.lib.update_unmerged(self.state)
+      end)
+    end
+
+    if lib.recent then
+      table.insert(refreshes, function()
+        logger.debug("[REPO]: Refreshing recent commits")
+        self.lib.update_recent(self.state)
+      end)
+    end
+
+    if lib.diffs then
+      local filter = (type(lib) == "table" and type(lib.diffs) == "table") and lib.diffs or nil
+
+      table.insert(refreshes, function()
+        logger.debug("[REPO]: Refreshing diffs")
+        self.lib.update_diffs(self.state, filter)
+      end)
+    end
+  else
+    logger.debug("[REPO]: Refreshing ALL")
+    self.lib.update_status(self.state)
+    a.util.scheduler()
+
+    for name, fn in pairs(self.lib) do
+      table.insert(refreshes, function()
+        logger.debug("[REPO]: Refreshing " .. name)
+        fn(self.state)
+      end)
+    end
+  end
+
+  logger.debug(string.format("[REPO]: Running %d refresh(es)", #refreshes))
+  a.util.join(refreshes)
```

Comment:
I have a wip timeout of individual refresh jobs, in case on hangs or doesn't complete.

Will rebase that ontop of this later.

I noticed that `merge.abort` can run indefinitely due to a race, that is why 😊

Reply:
Ah, very cool idea. Want to just do that via PR once this is merged to CKolkey? Rebasing means I have to fix the histories of all the subsequent branches 😟

Reply:
Ah no, I meant that I rebase my (yet to be commited on a suspended machine) changes on the ckolkey branch once all is wrapped up.

Opening a PR to ckolkey now would be a good idea, unless it makes the next few PRs conflict with their merges.

Reply:
Hmm. Maybe. But if you open it then I can at least have a look at what you're up to and plan accordingly :)